txtorcon/torcontrolprotocol.py

```diff
@@ -657,43 +675,45 @@ def _do_authenticate(self, protoinfo):
-                         len(self.cookie_data), "bytes")
-            self.client_nonce = os.urandom(32)
-
-            cmd = 'AUTHCHALLENGE SAFECOOKIE ' + \
-                  base64.b16encode(self.client_nonce)
-            d = self.queue_command(cmd)
-            d.addCallback(self._safecookie_authchallenge)
-            d.addCallback(self._bootstrap)
-            d.addErrback(self._auth_failed)
-            return
-
-        elif 'COOKIE' in methods:
-            cookie = re.search('COOKIEFILE="(.*)"', protoinfo).group(1)
-            with open(cookie, 'r') as cookiefile:
-                data = cookiefile.read()
-            if len(data) != 32:
-                raise RuntimeError(
-                    "Expected authentication cookie to be 32 "
-                    "bytes, got %d instead." % len(data)
-                )
-            txtorlog.msg("Using COOKIE authentication",
-                         cookie, len(data), "bytes")
-            d = self.authenticate(data)
-            d.addCallback(self._bootstrap)
-            d.addErrback(self._auth_failed)
-            return
-
-        if self.password_function:
+        if 'SAFECOOKIE' in methods or 'COOKIE' in methods:
+            cookiefile_match = re.search(r'COOKIEFILE="((?:[^"\\]|\\.)*)"', protoinfo)
+            if cookiefile_match:
+                cookiefile = cookiefile_match.group(1)
+                cookiefile = cookiefile.replace('\\\\', '\\')
+                cookiefile = cookiefile.replace('\\"', '"')
+                try:
+                    self._read_cookie(cookiefile)
+                except IOError as why:
+                    txtorlog.msg("Reading COOKIEFILE failed: " + str(why))
+                    cookie_auth = False
+                else:
+                    cookie_auth = True
+            else:
+                txtorlog.msg("Didn't get COOKIEFILE")
+
+        if cookie_auth:
+            if 'SAFECOOKIE' in methods:
+                txtorlog.msg("Using SAFECOOKIE authentication", cookiefile,
+                             len(self._cookie_data), "bytes")
+                self.client_nonce = os.urandom(32)
+
+                cmd = 'AUTHCHALLENGE SAFECOOKIE ' + \
+                      base64.b16encode(self.client_nonce)
+                d = self.queue_command(cmd)
+                d.addCallback(self._safecookie_authchallenge)
+                d.addCallback(self._bootstrap)
+                d.addErrback(self._auth_failed)
+                return
+
+            elif 'COOKIE' in methods:
+                txtorlog.msg("Using COOKIE authentication",
+                             cookiefile, len(self._cookie_data), "bytes")
+                d = self.authenticate(self._cookie_data)
+                d.addCallback(self._bootstrap)
+                d.addErrback(self._auth_failed)
+                return
+
+        if self.password_function and 'HASHEDPASSWORD' in methods or 'PASSWORD' in methods:
             d = defer.maybeDeferred(self.password_function)
```

Comment:
I like to put brackets even if they're the same as operator precedence, just to make it easier to read. So like ``if self.password_function and ('HASHEDPASSWORD' in methods or 'PASSWORD' in methods):``